### ISSUE TITLE
Enabling syslog audit backend results in HTTP response code 500 in response to `vault mounts`

### ISSUE THREAD

**Issue by jayed** [hashicorp/vault#3617]
**Environment:**

* Vault Version: Vault v0.8.3 ('6b29fb2b7f70ed538ee2b3c057335d706b6d4e36')
* Operating System/Architecture FreeBSD FreeBSD 11.1-STABLE #0 r325568 /usr/obj/usr/src/sys/64BIT  amd64

**Vault Config File:**
```
disable_mlock = "true"
default_lease_ttl = "4608h"
max_lease_ttl = "4608h"


listener "tcp" {
  address = "127.0.0.1:8200"
  tls_cert_file = "/usr/local/etc/ssl/vault.***.crt"
  tls_key_file = "/usr/local/etc/ssl/vault.***.key"
  tls_min_version = "tls12"
  tls_require_and_verify_client_cert = "false"
}

listener "tcp" {
  address = "99.100.100.227:8200"
  tls_cert_file = "/usr/local/etc/ssl/vault.***.crt"
  tls_key_file = "/usr/local/etc/ssl/vault.***.key"
  tls_min_version = "tls12"
  tls_client_ca_file = "/usr/local/etc/ssl/***.crt"
}

listener "tcp" {
  address = "[2600:1f1c:44f:1400:5f8c:1504:964d:61c3]:8200"
  tls_cert_file = "/usr/local/etc/ssl/vault.***.crt"
  tls_key_file = "/usr/local/etc/ssl/vault.***.key"
  tls_min_version = "tls12"
  tls_client_ca_file = "/usr/local/etc/ssl/***.crt"
}

storage "file" {
  path = "/usr/local/vault/data"
}
```

**Expected Behavior:**
`vault mounts` should return a list of mounts when the syslog audit mount enabled.

**Actual Behavior:**
Instead it returns:
```
Error reading mounts: Error making API request.

URL: GET https://127.0.0.1:8200/v1/sys/mounts
Code: 500. Errors:

* internal error
```

The audit log says:
```
Nov 27 17:21:00 ip-99-100-100-227 vault[55463]: {"time":"2017-11-27T17:21:00Z","type":"request","auth":{"client_token":"hmac-sha256:dc3ffc418ebcea5eb24ed92
646b64e839f6be8fca55018812120f873174f6b22","accessor":"hmac-sha256:3703e0ab7cb7735f4c3587f829e625f5c590a6d0f3b39dcdaf048f26980f9d5b","display_name":"root",
"policies":["root"],"metadata":null},"request":{"id":"276a5e33-5f24-78a4-e76e-69dfaa5cfce5","operation":"read","client_token":"hmac-sha256:dc3ffc418ebcea5e
b24ed92646b64e839f6be8fca55018812120f873174f6b22","client_token_accessor":"hmac-sha256:3703e0ab7cb7735f4c3587f829e625f5c590a6d0f3b39dcdaf048f26980f9d5b","p
ath":"sys/mounts","data":null,"remote_address":"127.0.0.1","wrap_ttl":0,"headers":{}},"error":""}
```

**Steps to Reproduce:**
`vault audit-enable syslog && vault mount`

**Important Factoids:**
* This machine has no swap.
* This worked in the past. It stopped after I mounted the syslog audit mountpoint.
```
vault audit-disable syslog
Successfully disabled audit backend 'syslog' if it was enabled

vault mounts
Path        Type       Accessor            Plugin  Default TTL  Max TTL  Force No Cache  Replication Behavior  Description
cubbyhole/  cubbyhole  cubbyhole_d3fe812d  n/a     n/a          n/a      false           local                 per-token private secret storage
secret/     kv         kv_fe4a90ef         n/a     system       system   false           replicated            key/value secret storage
sys/        system     system_358a875d     n/a     n/a          n/a      false           replicated            system endpoints used for control, policy and debugging
terraform/  kv         kv_910dbdd4         n/a     system       system   false           replicated
```

```
vault audit-enable syslog
Successfully enabled audit backend 'syslog' with path 'syslog'!

vault mounts
Error reading mounts: Error making API request.

URL: GET https://127.0.0.1:8200/v1/sys/mounts
Code: 500. Errors:

* internal error
```

**Comment by jefferai** [comment]
What does Vault's system log say? Internal errors are not returned to users (in case they are sensitive), but more information is logged into Vault's system log.

**Comment by jayed** [comment]
How do I generate a system log?

**Comment by jefferai** [comment]
Vault prints it to stderr when it's started.

**Comment by jayed** [comment]
```
vault mounts                        
Error reading mounts: Error making API request.                  

URL: GET https://127.0.0.1:8200/v1/sys/mounts
```

```2017/12/06 23:23:44.935857 [ERROR] audit: backend failed to log response: backend=syslog/ error=write unixgram ->/var/run/log: write: message too long             
2017/12/06 23:23:44.935900 [ERROR] core: failed to audit response: request_path=sys/mounts error=1 error occurred:                 

* no audit backend succeeded in logging the response             
2017/12/06 23:23:47.256799 [INFO ] expiration: revoked lease: lease_id=auth/token/create/f5ea536db741dbc60410fc51e9c7daca70a2dd43
```

```
**file log**                                                                                      
log: socket                                                                                                                      
**ls -al log**                                                                                    
srw-rw-rw-  1 root  wheel  0 Nov 21 01:38 log
**stat -x log**                                                                                   
  File: "log"                                                                                                                     
  Size: 0            FileType: Socket                                                                                             
  Mode: (0666/srw-rw-rw-)         Uid: (    0/    root)  Gid: (    0/   wheel)                                                    
Device: 227,862978275   Inode: 135    Links: 1                                                                                    
Access: Tue Nov 21 01:38:03 2017                                                                                                  
Modify: Tue Nov 21 01:38:03 2017                                                                                                  
Change: Tue Nov 21 01:38:03 2017 
```

`man 2 intro` makes it look like: **40 EMSGSIZE Message too long.  A message sent on a	socket was larger than the internal message buffer or some other network limit.**

A quick peek into `sysctl -a` makes it look like there's no tunable for the length of socket messages. I can't really tell more without digging into the libc code.

**Comment by jefferai** [comment]
Unfortunately I don't have a concrete answer... This isn't something we've seen before, and that includes people using the socket audit backed. It's possible it's something specific with the syslog daemon you're using or your system configuration. The only real suggestion I have right now is to use the file backend instead and have the logger ingest the entries.

**Comment by brianshumate** [comment]
I think this is actually from the [syslog protocol spec](https://tools.ietf.org/html/rfc5426#section-3.1) which requires that a single syslog message must be able to fit into a single UDP packet.

It seems that the options here are to either somehow reduce the audit log message size to fit into a single UDP packet or to enable a TCP listener on the syslog server if it's supported.

**Comment by jayed** [comment]
OK. Well, I'm not ready to visit "redoing my syslog infra" at the moment so I'll call this closed & revisit after I have time to tinker. Thanks!

**Comment by sbower** [comment]
I am also seeing this issues as our numbers of mounts have grown (~400).  We are getting the same error:

`{"timestamp":1529640151000,"message":"Jun 22 04:02:31 vault-lab-chs1-428b7604-us-east1-c-1 vault: 2018-06-22T04:02:31.753Z [ERROR] core: failed to audit response: request_path=sys/mounts error=\"1 error occurred:"}`

As in @jayed case disabling the audit backed alleviates the issues and  when the audit backend ids enabled again the error returns.

**Comment by sbower** [comment]
@jefferai could you reopen this ticket?  Or should I open a new ticket?

**Comment by jefferai** [comment]
@sbower See https://github.com/hashicorp/vault/issues/3617#issuecomment-351212237 -- I don't think there's anything to do here, other than add some documentation to the page warning users about UDP vs TCP.

**Comment by oxlay** [comment]
@jefferai I am confused about the doc mentioning UDP vs TCP for syslog.

Correct me if I am wrong, but I believe Vault is only using the Unix sockets like `/dev/log`: https://github.com/hashicorp/vault/blob/8575f8fedcf8f5a6eb2b4701cb527b99574b5286/vendor/github.com/hashicorp/go-syslog/builtin.go#L202

Indeed, Vault uses `NewLogger` https://github.com/hashicorp/vault/blob/8575f8fedcf8f5a6eb2b4701cb527b99574b5286/builtin/audit/syslog/backend.go#L67

which eventually leads to https://github.com/hashicorp/vault/blob/8575f8fedcf8f5a6eb2b4701cb527b99574b5286/vendor/github.com/hashicorp/go-syslog/builtin.go#L57

which has an empty string for `network` (first argument), so it's always using a Unix socket https://github.com/hashicorp/vault/blob/8575f8fedcf8f5a6eb2b4701cb527b99574b5286/vendor/github.com/hashicorp/go-syslog/builtin.go#L102-L103 

**So, regarding the issue above, I think the message just exceeds the maximum socket send buffer.**

Rather than suggesting to use TCP, the docs could maybe hint at http://man7.org/linux/man-pages/man7/socket.7.html, which suggests that increasing the value of `/proc/sys/net/core/wmem_default` could increase the send buffer size.

> SO_SNDBUF
              Sets or gets the maximum socket send buffer in bytes.  The
              kernel doubles this value (to allow space for bookkeeping
              overhead) when it is set using setsockopt(2), and this doubled
              value is returned by getsockopt(2).  The default value is set
              by the /proc/sys/net/core/wmem_default file and the maximum
              allowed value is set by the /proc/sys/net/core/wmem_max file.
              The minimum (doubled) value for this option is 2048.

**Comment by hmalphettes** [comment]
@oxlay thanks for your note. I observed the same thing. The fact that the current syslog is only able to send to send the messages on the unixsocket is also mentioned here: https://www.greenreedtech.com/vault-audit-logging/

I have dropped the syslog backend and I am only relying on a file now.
rotatelog is configured in this manner for me: https://github.com/hashicorp/vault/issues/472#issuecomment-432530492

I hope this helps.
